File: gc-starter/crud-spring-boot-starter/src/main/java/com/gc/starter/crud/advice/ParameterConvertRequestBodyAdvice.java
package com.gc.starter.crud.advice;

import com.gc.common.base.exception.IllegalAccessRuntimeException;
import com.gc.common.base.exception.IntrospectionRuntimeException;
import com.gc.common.base.exception.InvocationTargetRuntimeException;
import com.gc.common.base.utils.ReflectUtil;
import com.gc.starter.crud.query.QueryParameter;
import com.google.common.collect.Sets;
import lombok.SneakyThrows;
import org.springframework.core.MethodParameter;
import org.springframework.http.HttpInputMessage;
import org.springframework.http.converter.HttpMessageConverter;
import org.springframework.lang.NonNull;
import org.springframework.web.bind.annotation.ControllerAdvice;
import org.springframework.web.servlet.mvc.method.annotation.RequestBodyAdvice;

import java.beans.IntrospectionException;
import java.beans.PropertyDescriptor;
import java.lang.reflect.*;
import java.util.Set;

/**
 * 参数处理类，将map转为queryParameter
 * @author jackson
 * 2020/3/14 6:44 下午
 */
@ControllerAdvice
public class ParameterConvertRequestBodyAdvice implements RequestBodyAdvice {
    @Override
    public boolean supports(@NonNull MethodParameter methodParameter, @NonNull Type type, @NonNull Class<? extends HttpMessageConverter<?>> aClass) {
        return true;
    }

    @NonNull
    @Override
    public HttpInputMessage beforeBodyRead(@NonNull HttpInputMessage httpInputMessage, @NonNull MethodParameter methodParameter, @NonNull Type type, @NonNull Class<? extends HttpMessageConverter<?>> aClass) {
        return httpInputMessage;
    }

    @SneakyThrows
    @NonNull
    @Override
    public Object afterBodyRead(@NonNull Object body, @NonNull HttpInputMessage httpInputMessage, @NonNull MethodParameter methodParameter, @NonNull Type type, @NonNull Class<? extends HttpMessageConverter<?>> aClass) {
        if (type instanceof ParameterizedType) {
            type = ((ParameterizedType) type).getRawType();
        }
        final Class<?> clazz;
        if (type instanceof Class) {
            clazz = (Class<?>) type;
        } else {
            clazz = Class.forName(type.getTypeName());
        }
        if (body instanceof QueryParameter) {
            this.convertParameter((QueryParameter<?, ?>) body, clazz);
        }
        return body;
    }

    @Override
    public Object handleEmptyBody(Object body, @NonNull HttpInputMessage httpInputMessage, @NonNull MethodParameter methodParameter, @NonNull Type type, @NonNull Class<? extends HttpMessageConverter<?>> aClass) {
        return body;
    }

    /**
     * 转换参数
     * @param parameter 参数
     */
    @SuppressWarnings("rawtypes")
    private void convertParameter(QueryParameter parameter, Class clazz) {
        // 处理PageQueryParameter
        this.addParameter(parameter, clazz);
    }

    /**
     * 添加参数
     * @param parameter 参数
     * @param clazz 实体类类型
     */
    @SuppressWarnings("rawtypes")
    private void addParameter(QueryParameter parameter, Class<? extends QueryParameter> clazz) {
        final Set<Field> fieldSet = Sets.newHashSet();
        ReflectUtil.getAllFields(clazz, fieldSet);
        try {
            for (Field field : fieldSet) {
                if (!Modifier.isFinal(field.getModifiers()) && !Modifier.isStatic(field.getModifiers())) {
                    String name = field.getName();
                    if (parameter.containsKey(name)) {
                        Object value = parameter.get(name);
                        parameter.remove(name);
                        PropertyDescriptor propertyDescriptor = new PropertyDescriptor(name, clazz);
                        propertyDescriptor.getWriteMethod().invoke(parameter, value);
                    }
                }
            }
        } catch (IllegalAccessException e) {
           throw new IllegalAccessRuntimeException(e);
        } catch (InvocationTargetException e) {
            throw new InvocationTargetRuntimeException(e);
        } catch (IntrospectionException e) {
            throw new IntrospectionRuntimeException(e);
        }
    }

}

File: smart-starter/crud-spring-boot-starter/src/main/kotlin/com/smart/starter/crud/advice/ParameterConvertRequestBodyAdvice.kt
package com.smart.starter.crud.advice

import com.smart.common.utils.DateUtil
import com.smart.common.utils.JsonUtil
import com.smart.starter.crud.query.QueryParameter
import org.springframework.core.MethodParameter
import org.springframework.http.HttpInputMessage
import org.springframework.http.converter.HttpMessageConverter
import org.springframework.web.bind.annotation.ControllerAdvice
import org.springframework.web.servlet.mvc.method.annotation.RequestBodyAdvice
import java.lang.reflect.Type
import java.util.*
import kotlin.reflect.KMutableProperty1
import kotlin.reflect.full.memberProperties
import kotlin.reflect.jvm.javaField

/**
 * 参数转换器
 * @author ming
 * 2019/9/13 下午10:12
 */
@ControllerAdvice
class ParameterConvertRequestBodyAdvice : RequestBodyAdvice {

    companion object {
        private val excludeProperties = listOf("entries", "keys", "size", "values")
    }

    override fun afterBodyRead(body: Any, inputMessage: HttpInputMessage, parameter: MethodParameter, targetType: Type, converterType: Class<out HttpMessageConverter<*>>): Any {
        if (body is QueryParameter) {
            this.convertBody(body)
        }
        return body
    }

    override fun beforeBodyRead(inputMessage: HttpInputMessage, parameter: MethodParameter, targetType: Type, converterType: Class<out HttpMessageConverter<*>>): HttpInputMessage {
        return inputMessage
    }

    override fun handleEmptyBody(body: Any?, inputMessage: HttpInputMessage, parameter: MethodParameter, targetType: Type, converterType: Class<out HttpMessageConverter<*>>): Any? {
        return body
    }

    override fun supports(methodParameter: MethodParameter, targetType: Type, converterType: Class<out HttpMessageConverter<*>>): Boolean {
        return true
    }

    /**
     * 转换数据
     */
    @Suppress("UNCHECKED_CAST")
    private fun convertBody(data: QueryParameter) {
        // 获取对象的所有树形
        val properties = data :: class.memberProperties
        properties.forEach { property ->
            val name = property.name
            val value = data[name]
            if (!excludeProperties.contains(name) && value != null) {
                val propertyType = property.javaField!!.type
                if (property is KMutableProperty1) {
                    property as KMutableProperty1<QueryParameter, Any?>
                    when {
                        value is Map<*, *> -> property.set(data, JsonUtil.toJavaObject(value, propertyType))
                        propertyType == Date :: class.java -> // 处理时间类型
                            property.set(data, DateUtil.convertDate(value.toString()))
                        else -> property.set(data, value)
                    }
                    data.remove(name)
                }
            }
        }
    }
}